up-transport-vsomeip/src/transport_inner/transport_inner_handle.rs

```diff
@@ -0,0 +1,880 @@
+            local_authority_name.clone(),
+            remote_authority_name.clone(),
+            ue_id,
+        ));
+
+        let engine = UPTransportVsomeipInnerEngine::new(ue_id, Some(config_path));
+        let point_to_point_listener = TimedStdRwLock::new(None);
+        let config_path = Some(config_path.to_path_buf());
+
+        Ok(Self {
+            engine,
+            storage,
+            point_to_point_listener,
+            config_path,
+        })
+    }
+
+    #[cfg(test)]
+    pub(crate) fn new_supply_storage(
+        storage: Arc<dyn UPTransportVsomeipStorage>,
+    ) -> Result<Self, UStatus> {
+        let engine = UPTransportVsomeipInnerEngine::new(storage.get_ue_id(), None);
+        let point_to_point_listener = TimedStdRwLock::new(None);
+        let config_path = None;
+
+        Ok(Self {
+            engine,
+            storage,
+            point_to_point_listener,
+            config_path,
+        })
+    }
+
+    #[cfg(test)]
+    pub(crate) fn new_with_config_supply_storage(
+        config_path: &Path,
+        storage: Arc<dyn UPTransportVsomeipStorage>,
+    ) -> Result<Self, UStatus> {
+        let engine = UPTransportVsomeipInnerEngine::new(storage.get_ue_id(), Some(config_path));
+        let point_to_point_listener = TimedStdRwLock::new(None);
+        let config_path = Some(config_path.to_path_buf());
+
+        Ok(Self {
+            engine,
+            storage,
+            point_to_point_listener,
+            config_path,
+        })
+    }
+
+    async fn await_engine(
+        function_id: &str,
+        rx: oneshot::Receiver<Result<(), UStatus>>,
+    ) -> Result<(), UStatus> {
+        match timeout(Duration::from_secs(INTERNAL_FUNCTION_TIMEOUT), rx).await {
+            Ok(Ok(result)) => result,
+            Ok(Err(_)) => Err(UStatus::fail_with_code(
+                UCode::INTERNAL,
+                format!(
+                    "Unable to receive status back from internal function: {}",
+                    function_id
+                ),
+            )),
+            Err(_) => Err(UStatus::fail_with_code(
+                UCode::DEADLINE_EXCEEDED,
+                format!(
+                    "Unable to receive status back from internal function: {} within {} second window.",
+                    function_id, INTERNAL_FUNCTION_TIMEOUT
+                ),
+            )),
+        }
+    }
+
+    async fn send_to_engine_with_status(
+        tx: &Sender<TransportCommand>,
+        transport_command: TransportCommand,
+    ) -> Result<(), UStatus> {
+        tx.send(transport_command).await.map_err(|e| {
+            UStatus::fail_with_code(
+                UCode::INTERNAL,
+                format!(
+                    "Unable to transmit request to internal vsomeip application handler, err: {:?}",
+                    e
+                ),
+            )
+        })
+    }
+
+    fn get_storage(&self) -> Arc<dyn UPTransportVsomeipStorage> {
+        self.storage.clone()
```

Comment:
Doing an audit of the code just now, this does not appear to be on the hot path, since it's used only in sections of code related to registering, unregistering, and the Drop impl.

Could you go into more detail on what makes it feel awkward? Performing many potential clones on an `Arc`?

Reply:
Yes, but I guess the real culprit it that you need/want to expose the inner structure by handing out the storage in the first place ...

Reply:
Hmm, okay. I'll think on this. :thinking:

Reply:
I removed the get_storage() method. Wasn't needed anymore.